[MMEEvent] 1.0 Beta (Refactor)

dmiluski opened the pull request:
Goal:
Verbose constructors allow for easy testability as well as helpful interfaces.

Expectation:
Event creation should support determinant outcomes. EventManager should ONLY need to focus on managing queues/event routing rather than performing Event Building requirements.

Approach:
1. To support determinant outcomes, request all information at the time of initializer to validate dictionary structure outcome. This allows testing against normally dynamic content such as UIApplication state or dates.
2. When we expect a given date format, we can accept a date argument and format it ourselves. This also prevents invalid data such as raw strings making it directly to the dictionary (would be strong in seprate classes rather than falling through to designated init with dictionary.

Changes:
+ Add examples via mapLoadEvent, mapTapEvent, mapDragEvent, locationEvent, Turnstile, Visit event
+ Add similar tests verifying the input/output is as expected.
+ Migrate turnstile to convenience initializer which performs event building in the model rather than the manager.
- Remove event building specific testing from events manager tests

dmiluski committed:
[MMEEvent] Strong Convenience Initializers

Goal:
Verbose constructors allow for easy testability as well as helpful interfaces.

Expectation:
Event creation should support determinant outcomes. EventManager should ONLY need to focus on managing queues/event routing rather than performing Event Building requirements.

Approach:
1. To support determinant outcomes, request all information at the time of initializer to validate dictionary structure outcome. This allows testing against normally dynamic content such as UIApplication state or dates.
2. When we expect a given date format, we can accept a date argument and format it ourselves. This also prevents invalid data such as raw strings making it directly to the dictionary (would be strong in seprate classes rather than falling through to designated init with dictionary.

Changes:
+ Add examples via mapLoadEvent, mapTapEvent, mapDragEvent, locationEvent, Turnstile, Visit event
+ Add similar tests verifying the input/output is as expected.
+ Migrate turnstile to convenience initializer which performs event building in the model rather than the manager.
- Remove event building specific testing from events manager tests

dmiluski committed:
[CI] Carthage CI Fail

Problem:
Bitrise Carthage step fails if it is run with an empty resolved file.

Change:
Retain this resolve file contents until the carthage dependencies are merged to master.

dmiluski committed:
[Warning] Missing import

dmiluski committed:
[CLLocationManager Enum] Test CL Category

Goal:
Bump Test numbers by testing this output

Approach:
Provide testable evaluation by providing CLLocationManager subclasses with prespecified statuses.

Changes:
+ Add Test Class
+ Add Mock Classes
+ Udpate category to pull from self allowing for mock subclasses

dmiluski committed:
[UIKit Extensions] Testing

Goal:
Add testing to UIKit category methods

Approach:
Leverage subclasses where possible to stub responses. For classes which do not allow for subclassing, leverage helper functions that take trait/arguments that do allow for testing.

Changes:
+ Add Mock Classes for testing
+ Add Tests for UIDevice/NSExtension/UIApplication Category Tests

dmiluski committed:
[APIClient] Remove Hooks from Init

Problem:
While lookig into EventsManager testing additions, found that Client init function was overly complicated. Although passing in the listener closures is nice for ensuring it’s setup at the beggining, it needlessly complicates test cases that don’t need to worry about them.

Approach:
Instead of passing on init, allow owners to register listeners and call.

Changes:
+ Add Register Methods for adding hook handlers
+ Set default arrays, and call listeners on events
+ Update calling points

Notes:
Returned count value from batch to events until I get confirmation on what value should be reported?

alfwatt reviewed (changes requested):
I think most of these can be moved into private headers since we fill out the fields on Maps behalf at least, already.

The date parameter versions make total sense for testing but in production we should always set the date to 'now' on event creation. There should be a single, simple no-argument create method that's publicly exposed, and the rest can live in the private header for use by our unit and integration tests.

It would be better in production to have detailed NSError objects which we can report back, in place of log messages which are convenient for development and testing but end up being hard to get for diagnostic purposes.

dmiluski commented on MapboxMobileEvents/MMEEvent.h:
```diff
@@ -1,10 +1,12 @@
 @import Foundation;
 
 #import <MapboxMobileEvents/MMETypes.h>
+#import "MMEEventConfigProviding.h"
```
<img width="839" alt="Screen Shot 2020-05-26 at 12 32 14 PM" src="https://user-images.githubusercontent.com/5083390/82942539-f54cbf80-9f4c-11ea-95da-f08f67d1f9cc.png">

dmiluski committed:
Events Convenience Initializers

Goal:
Although the DI param allows for testing, when actually called, SDKs would prefer fewer arguments if they can be inferred.

Changes:
+ Include convenience helper with no args.

dmiluski committed:
[Spelling] MMEEventConfigProviding

MMEEventCOnfigProviding -> MMEEventConfigProviding

dmiluski commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -83,20 +86,114 @@ + (instancetype)eventWithAttributes:(NSDictionary *)attributes error:(NSError **
-    return [self eventWithAttributes:eventAttributes];
-}
+    if (!config.accessToken) {
+        MMELog(MMELogInfo, MMEDebugEventTypeTurnstileFailed, ([NSString stringWithFormat:@"No access token sent - can not send turntile event, instance: %@",
```
That's definitely possible and would provide insight into internals.

dmiluski replied:
Updated to include error feedback mechanism in init

dmiluski committed:
[Project File Mishap] Mock File incorrectly in the wrong target

Problem:
While looking into code coverage, noticed we were testing a mock providing file.

Change:
+ Move to test target

dmiluski committed:
[Turnstile] Init Error Reporting

Goal:
Although console logs are beneficial, we are currently not getting feedback of how many applications/sdks may be running into a turnstile init error.

To gather feedback should this ever happen, provide error ref feedback during init, allowing for reporting.

Approach:
Include turnstile failure init reporting via error reference. The side effect of this is now Turnstile tests result in a new error event being added to the queue which requires tests to be update to go from no queued/posted items to queued error events.

Changes:
+ Add error reference to be filled on Turnstile init
+ Update tests now that this leads to a new error MMEEvent. being queued in the regular event queue.

dmiluski committed:
[NSUserDefaults] Removed Store from Interface

Problems:
Although NSUserDefaults is fine as a datastore for key/value pairs, leveraging it as the interface can cause a number of problems.

1. Couples with NSUserdDefaults preventing the migration elsewhere should it be necessary.

2. Requires use of mme_config while the rest of the accessors are able to be accessed on any NSUserDefault instance allowing for potential misuse on accident.

eg. NSUserDefaults.shared vs NSUserDefaults.mme_config

3. Given the register defaults as well as writing to the same location, we lose the ability to determine if the value is: Default, Set by a developer, or coming from a remote config. This is where the abstraction will come in nicely in the future to provide this feedback given some GUI settings adjuster.

4. Given Preferences exists, having two sources of truth would be more confusing than beneficial.

Approach:
1. Removed NSUserDefaults+MMEConfiguration implementation details other than items like suite setup which will still prove useful for internal use.
 2. Adjust Preferences individual components to no longer depend on registering defaults. Instead, calculate defaults so resets don’t require multiple steps.

Changes:
+ Migrate remaining register defaults to be calculated on property accessor ensuring availability of value regardless if someone delete/removes it. Leverage objects rather than bool/double to check if item exists rather than defaulting to 0/false these types
+ Add additional
+ Remove imports that don’t need it
+ Expose ClientID getter as it exists but didn’t notice it was required until seeing MacOS ifdef

dmiluski committed:
[MMEEventConfigProviding] Move Public -> Internal

Problem:
MMEEVentConfigProviding was originally puposed to enable read only access of configurations through the project to enable dependency injection of instance dependencies instead of singletons allowing for determinant testing.

Although it’s served it’s purpose it doesn’t need to be exposed publically since consuming SDKs can interact with preferences as they may rightly want to do with a GUI editor.

Expectations of Success:
1. Remove MMEEventConfigProviding from public header space
2. SDK Builds

Approach:
To remove MMEEventConfig from the public header space it needs to be removed from MMEEvent.h imports. To accomplish this, the only model which requires this import (Turnstile) was moved to an internal header (along with designated initializers that don’t need to be exposed to the public)

Changes:
- Remove MMEEventConfig from Public header space
+ Migrate MMEEvent initializers to a private header to allow access internally, but limit visibility externally
+ Update comments to follow objective-c style docs

dmiluski committed:
[XCode Project] Xcode Suggested Cleanup

Duplicate Copy File

Xcode offered to autofix

dmiluski committed:
[Test Target] Host App Depends on SDK

Problem:
When running Test, devleoper is required to manually build the framework first, then run tests.

Approach:
Current Project linking doesn’t know this.

Change:
+ Add TestHost dependency for Framework

dmiluski committed:
[MMELocationManager] Tests

Goal:
Migrate suite of commented out Cedar tests to XCTestCases

Approach:
1. Add LocationManagerCallCounter to evaluate how often calls are made as expected to locationManager from MMELocationManager.
2. Add Documentation to MMELocatioNmanager methods and delegates.

Changes:
+ Migrate LocationManager protocol to its own separate file and rename locationManager to limit confusion between protocol and class.
+ Add LocationManagerCallCounter to provide call counting for LocationManager calls. Include specialized subclasses to provide various Authorization states.
+ Migrate Cedar tests to XCTestCase. Include additional coverage to boost file functional testing to >~95%

dmiluski committed:
[XcodeProject Warning] Missing Protocol Access

Problem:
Tests were unable to see protocol definition

Change:
Include protocol in accessible area for tests

dmiluski committed:
[EventsManager] Repeated Starts Over Observing

Problem:
In the event of multiple observation calls, (eg several SDKs included and each starting the Telemetry SDK.

Discovered based on inspection, but reproduced via
```
    [MMEEventsManager.sharedManager startEventsManagerWithToken:@"Floof"];

    // Loop Start it to see what happenes
    [NSTimer scheduledTimerWithTimeInterval:3 repeats:YES block:^(NSTimer * _Nonnull timer) {
        [MMEEventsManager.sharedManager startEventsManagerWithToken:@"Floof"];
    }];
```
Where upon putting application in the background, the Events Manager receieved the number of notifications as the timer had looped. So for 3 sdks interacting with Eventsmanager, EventsManager may get 3 notification calls.

Expectations:
For each repetetive start(:::) call, Events manager observer selectors should only be called on each triggering notification name.

Note:
Is this something that we should even allow? Or perhaps only dispatch once?

dmiluski committed:
[MMEEventsManager] Client/ConfigService Setup Test

Goal:
Ensure Client and Config service are configured as expecrted.

Expectations:
Provide testing strategy to ensure Listener handlers behave as expected

Outcome:
1. While applying tests, determined that reportSentBytes was not acting as expected. Instead of reporting all sent bytes, it was only reporting as a reaction for event count updates resulting in a lack of information coming from other API calls.
2. While applying tests, discovered we were reporting the full event size rather than the batch size, potentially resulting in reporting larger numbers for success/failure than expected.
3. Noticed that postMetadata failed to report count due to a missed migration. Fixed.

Approach:
1. Extract Client/Config service setup from start(:::) into their own methods which can be called to verify expected setup responsibilities.
2. Verify each event is registered on configuration
3. Verify those registered block calls result in expected handling.

Changes:
+ Add Docs to touched methods
+ Extract configureClient and config creation to it’s own separate methods, expose in Private header.
+ Add MetricsManagerCallCounter to verify calls into class. May be ideal in the future to verify contents rather than just call counts. Eg. an array of arguments.
+ Add Tests to verify methods registered listeners as expected
+ Add Tests to verify registered listeners reacted to calls by verifying their outcomes on trigger.
+ Updated test errors to use specific error domain/code for use in extracting specific userInfo for report

dmiluski committed:
[MMEUIApplicationWrapper] Tests

Goal:
Boost class coverage to 100%

Approach:
Rather than directly accessing UIApplication.shared, provide mechanism to inject dependency allowing to test this classes responsibilities.

To achieve this, pass in a MMEUIApplicationWrapper, the interface that is captured by this class and the UIApplication. By passing this in, we can ensure our wrapper is making the correct calls to the dependency UIApplication.

Change:
+ Add Docs
+ Add dependency property
+ Add Designated/Default initializers
+ Add Tests to verify calls/responses

dmiluski committed:
[Consumption Compatibility]

Problem:
Existing Consumers expect some items currently not exposed via SDK.

What:
- NSUserDefaults+MMEConfiguration
- pauseOrResumeMetricsCollectionIfRequired

Change:
- Include these for now

Todo:
reach out to Maps team to see if these are still required?

dmiluski committed:
[MMELogLevel] - Log Level Setter

Goal:
Maps is currently reading/setting logging behaviors from bundle via a private class extension. This is not ideal behavior as this can lead to crashes should internal method names change.

Approach:
Although SDK automatically reads these same key/values from bundle, it may also be ideal to allow for override of log levels at run time (perhaps via a debug menu)

Changes:
+ Expose public log level

dmiluski committed:
[MMEEventsManager] Additional Tests

Goal:
Boost overall test numbers

Approach:
A number of methods embedded functionality in completion handlers making testing more difficult. By extracting the completion handlers to it’s own function, the responsibilities of the completion handler can be tested

Alternative Option:
Add a completion handler to the existing method to evaluate the same behaviros after executing. Although possible, it required adjusting a public interface to enable testing as well as requiring additional infrustructure setup to mock leaving the extracted function as the preferred approach.

To add more test coverage I first focused on the functions which were not being tested such as delegate calling, and completion handlers.

Changes:
+ Add TestCase as delegate for monitoring delegate calls
+ Compare Delegate calls to verify they are being made
+ Extract completion handlers to their own methods which can be artificially called to verify their behaviors.

alfwatt commented on MapboxMobileEvents.xcodeproj/project.pbxproj:
```diff
@@ -221,8 +221,8 @@
 		F47265692466872200BD7055 /* MMEConfigServiceTests.m in Sources */ = {isa = PBXBuildFile; fileRef = F47265682466872200BD7055 /* MMEConfigServiceTests.m */; };
-		F4913F452453610800C28338 /* MMEEventConfigProviding.h in Headers */ = {isa = PBXBuildFile; fileRef = F4913F442453610800C28338 /* MMEEventConfigProviding.h */; };
-		F4913F462453610800C28338 /* MMEEventConfigProviding.h in Headers */ = {isa = PBXBuildFile; fileRef = F4913F442453610800C28338 /* MMEEventConfigProviding.h */; };
+		F4913F452453610800C28338 /* MMEEventConfigProviding.h in Headers */ = {isa = PBXBuildFile; fileRef = F4913F442453610800C28338 /* MMEEventConfigProviding.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Can we make this protocol 'MMEConfigurationProviding'?

dmiluski replied:
Updated

alfwatt commented on MapboxMobileEvents/MMEPreferences.h:
```diff
@@ -62,6 +62,9 @@ NS_ASSUME_NONNULL_BEGIN
 /// Host SDK Version
 @property (nullable, nonatomic, copy) NSString *legacyHostSDKVersion;
 
+/// Unique Identifier for the client
```
Guess I can only add comments to changed lines…

Please rename this class `MMEConfigation` and group it with `MMEConfigurationProviding` and `NSUserDefault+MMEConfiguration` files

dmiluski replied:
Updated

alfwatt commented on MapboxMobileEvents/Protocols/MMEEventConfigProviding.h:
```diff
@@ -61,6 +61,9 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, copy, readonly) NSString *legacyUserAgentString;
 
+/// Unique Identifier for the client
+@property (nonatomic, copy, readonly) NSString *clientId;
```
This should be `configuationId`

dmiluski replied:
Could you elaborate on this?

This particular change was a port from:
https://github.com/mapbox/mapbox-events-ios/commit/48145e285c68ac9f5323a1232f3c20f7ba1aee46

In which the cid: Client id was documented in the string constant.

I was wondering if I also need to update documentation for what this stands for as reading through code, I'm not sure what it represents? The unique instance of a MME SDK in an app? Something else?

As It looks to only be used internally for:
- Vendor id if used on platform != (iOS | tvOS)

dmiluski committed:
[PR Feedback] Rename MMEConfigurationProviding

Goal:
Rename MMEEventConfigProviding -> MMEConfigurationProviding

Why:
Event is redundant in the sdk and the configurations are not necessarily tied to event only behaviors.

Changes:
Xcode refactor name changed in project

dmiluski committed:
[PR Feedback] Rename MMEConfiguration

Goal:
Rename MMEPreferences -> MMEConfiguration

Context:
Previously named preferences as it was a mix of values provided in Bundle/NSUserDefaults, as well as some which came remotely from APIEventConfig service.

Why:
Preferences when used with property names config causes a disconnect in what property/values will be access. To remedy this confusion, rename to MMEConfiguration.

Changes:
Xcode refactor name changed in project

MMEPreferencesTests -> MMEConfigurationTests

dmiluski committed:
[PR Feedback] Internal -> Private

Goal:
Match existing category patterns of private instead of internal

Changes:
Xcode refactor renamed file/imports/category

dmiluski committed:
[PR Feedback] Move LogLevel Enum to Constants

Goal:
Limit file imports for Public consumption

To limit the need to expose another class to consumers, this enum definition can be placed in constants with other enums for general purpose use.

Changes:
- Delete MMELogLevel.h
+ Move contents to Constants.h
+ Update imports

dmiluski commented on MapboxMobileEvents/MMEEvent+Internal.h:
```diff
@@ -0,0 +1,111 @@
+NS_ASSUME_NONNULL_BEGIN
+
+// MARK: - Strong Event Models via Convenience Initializers
+@interface MMEEvent (Internal)
```
Updated

dmiluski commented on MapboxMobileEvents/MMELogLevel.h:
```diff
@@ -0,0 +1,26 @@
+/*! Log Levels for Runtime Filtering */
+typedef NS_ENUM(NSUInteger, MMELogLevel) {
```
Migrated/updated

dmiluski committed:
[MMELog] Line Count Tweaks

Goal:
For macros, the newline doesn’t quite help readability given the lack of param/type safety.

These newlines added extra lines missing code coverage that couldn’t be validated unless we injected an MMELog instance rather than th macro
